Use dual quaternion skinning for Skeletons

beiller opened the pull request:
Work in progress to use dual quaternions for Skeleton3D

TODO:
- Make it toggleable
- Errors can occur when rotations go very extreme, then back to normal. They get stuck in one particular orientation due to the `avoid_jumps` function, but that function prevents small rotations from suddenly flipping
- Applying local non-uniform scale is not working as it did with normal matrix transforms and I'm not sure why yet.

---

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/9235*

beiller committed:
Initial dual quaternion implementation

beiller committed:
mostly working

AThousandShips commented:
Thank you for your contribution!

You should target `master` not `4.2`, changes can then be cherry picked, see [here](https://docs.godotengine.org/en/latest/contributing/documentation/contributing_to_the_documentation.html#contributing-changes).

AThousandShips commented on scene/3d/skeleton_3d.cpp:
```diff
@@ -224,6 +224,61 @@ void Skeleton3D::_update_process_order() {
 
+Transform3D dual_quaternion_to_matrix(Quaternion Qn, Quaternion Qd)
+{
+	//Original version of this function is in https://users.cs.utah.edu/~ladislav/dq/dqs.cg
```
You need to add the license note from the original code

beiller replied:
Hey so this is just a math reference. Do we think we need copyright? But actually going to that site I realize now it has a function we can use, the scaling which may fix my TODO item!

AThousandShips replied:
Did you base your code on that? You say "original version" so that implies it's copied or based on it

beiller replied:
I actually based my code on someone else's GDscript code (https://godotengine.org/asset-library/asset/2062). The reference code is a cg shader not even the same language. But I am curious what we should do here. I have no problem the copyright notice would look something like this:

`Copyright (C) 2006-2007 University of Dublin, Trinity College, All Rights Reserved`

But really I wonder is it necessary since only a part of that code is "referenced" and the reference is just a math formula (convert dual quaternion to mat4).

This comment is actually from the GDScript code and I just carried it over.

beiller replied:
FYI its stated in the notice at the top of the link:
1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, **an acknowledgment in the product documentation would be
     appreciated but is not required.**

AThousandShips replied:
That's for *using* the library though, for *copying* the source it says "This notice may not be removed or altered from any source distribution."

akien-mga replied:
I had a look, I can find a number of papers with similar formula, and the rotational part is also on Wikipedia: https://en.wikipedia.org/wiki/Dual_quaternion#Dual_quaternions_and_4%C3%974_homogeneous_transforms

Overall my layman understanding is that math formulas aren't copyrightable, since they're facts (math proofs could be copyrighted though). So I think we don't need to include the full license statement for borrowing this formula to convert a dual quaternion to a 4x3 matrix.

I would keep the link to the original source though, which is useful to check assumptions and double check that the code was correctly ported.

AThousandShips commented on servers/rendering/renderer_rd/shaders/skeleton.glsl:
```diff
@@ -113,6 +113,58 @@ uint encode_tang_to_uint_oct(vec4 base) {
+
+mat4 dual_quaternion_to_matrix(vec4 Qn, vec4 Qd)
+{
+	//Original version of this function is in https://users.cs.utah.edu/~ladislav/dq/dqs.cg
```
Same here

beiller commented:
Updated to point to master branch

Calinou commented:
Thanks for opening a pull request :slightly_smiling_face: 

Feature pull requests should be associated to a *feature proposal* to justify the need for implementing the feature in Godot core. Please open a proposal on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository (and link to this pull request in the proposal's body).

beiller committed:
Try per-vertex antipodality handling

beiller committed:
Blend between linear and quaternion as well as fix stability

beiller committed:
Clean up unused code

beiller committed:
Fix signatures

beiller committed:
Need to get the quaternions

etodd committed:
Fix particle jitter when scene tree is paused.

(cherry picked from commit 87efa4d210e1bdd0cec69418e8242edc16855aa7)

aryan-11825114 committed:
fix changed signal emission in Curve::set_point_offset

(cherry picked from commit cd71f97209b5a1512e0620779d0643a724177a5c)

CreatedBySeb committed:
Fix a segfault in ConnectDialog if the destination node does not exist

(cherry picked from commit ed108fc9939a1246a028679e764cacb6d5498306)

raulsntos committed:
[.NET] Use `ObjectID` when converting `Variant` to `GodotObject`

(cherry picked from commit bfcc389e21dcb8426ca4f1ba7efb2debf5139ec5)

ryevdokimov committed:
Fix spinbox decimal issues when `update_on_text_changed` = true

(cherry picked from commit d8318deb0a53706f5c039bf2a01f24a84e7da07e)

beiller committed:
checkpoint

arkology committed:
Use `FlowContainer` for `EditorNetworkProfiler` bar

(cherry picked from commit c7ef85a92fcf283e95438c86210118c94ed47de7)

darksylinc committed:
Add ASTC HDR format variants

(cherry picked from commit fcd785ace2aedb0e1530833df099d5d437bd2e33)

dsnopek committed:
JavaClassWrapper: Improve handling of typed array arguments

(cherry picked from commit d7672128b5096ba23b3900da335b6ce8cfcfd2c3)

berarma committed:
Theora: Fix YUV422/444 to RGB. Fixes #87705

The last line wasn't being converted thus was missing in the resulting frame.

(cherry picked from commit 0882702827075c5b091bdad20ed0d73de2622c22)

Giganzo committed:
Fix Tree keyboard navigation in RTL direction

(cherry picked from commit bd6312383549211b5123cad2194ec8fbdd5d0458)

mooflu committed:
Apply fix_alpha_edges for both theme icons and font glyphs.
Only apply fix_alpha_edges for SVG glyphs.
Allow modulate for svg glyphs.

(cherry picked from commit 70a8421c307a660e43865e0e58c9f232a33c3d96)

orange-anvil-games committed:
fix voxelizer normals

(cherry picked from commit 0c57232f78c7c23491eea93e9d73ddf866bfa3e1)

kleonc committed:
Fix 2D quad primitive missing lighting data in GLES3 renderer

(cherry picked from commit b3769129cbab1cde77265664f0c55d78df2622bc)

darksylinc committed:
Fix uninitialized value in Tonemap

And anything that uses luminance.

The class Luminance in luminance.cpp is in charge of averaging the
luminance of all pixels.
It performs multiple passes until it reaches a 1x1 texture containing
the total average. This is standard luminance averaging on GPU.

Then the "result" of this frame and the "prev_frame_result" are averaged
together at a certain speed to mimic eye adaptation.

Then this avarege becomes the "source" for the next frame. This is done
here:

```cpp
SWAP(p_luminance_buffers->current,
p_luminance_buffers->reduce.write[p_luminance_buffers->reduce.size() -
1]);
```

So far pretty normal stuff.

**The problem is: prev_frame_result IS UNINITIALIZED**. Therefore it's
possible for prev_frame_result to contain garbage values like -5+e15
which causes the screen to stay black for a minute until eye adaptation
catches up.

Windows will always force allocations to be reset to 0, but Linux does
not do that.
However Windows just delays the bug; because it's possible for VMA to
reuse a block.

You can repro this bug by downloading Bistro, creating a camera,
selecting a default scene; and then launching Bistro.

Everything will work fine.
Until you decide to resize the window. It takes a few tries on Godot,
but eventually the screen becomes black.

If you wait around a minute, the screen will "unblack" itself back to
normal.

Even if it's not stuck in black after resize, you may notice that every
resize is inconsistent in how the eye adaptation catches up (i.e.
sometimes it flashes to white, sometimes it does not).

If you can't repro the bug, you need to try harder by doing arbitrary
resizes until it triggers.
Also, I advise to try this on Linux; since Windows' sanitization of
memory gets in the way.

There's probably multiple tickets already filled around issues that were
rooted in luminance calculations starting from uninitialized memory.

This PR sets a default value of 0, which causes the screen to always
flash white after resize. Setting a different value like 0.1 makes the
flash effect weaker. Setting it to a high value like 5.0 makes the
screen flash from dark instead.

This bug can be backported to 4.3. I don't know if it can be backported
to earlier; as the render graph makes sure the texture_clear() calls
gets issued in the right place; whereas in <= 4.2 it might be
problematic depending on when Luminance::LuminanceBuffers::configure is
being called.

(cherry picked from commit 8888f9e649a37d00b65996cbe8a56e42bee6647c)

OhiraKyou committed:
Fix GPU particles not emitting at some amounts when scale curve is zero

(cherry picked from commit 8df75953ec417454b6e056137d3262b7cc948bfd)

etherealxx committed:
Add "Go Online" button on Export Template Manager

Godot 4.4 introduces network mode, which by default sets to offline.
Some features are disabled on offline mode, including downloading export
templates. Newcomers to the engine that has no knowledge about the network
mode might be confused on why the Export Template Manager tells them that
they're offline, despite them having internet connection.

This commit introduces a message that tells the user that online mode is
required, and a link button that user can click to enable online mode from
the Export Template Manager popup UI. Some code changes also made to make
sure that the button and message only shows on official build only.

(cherry picked from commit 9cbfeff940cf131376d2c5c6f2916caf8865b23a)

marcgpuig committed:
Disabled depth test for line and handle materials

(cherry picked from commit 57d8ef5d2343c669715798345a84d32d91bc0c4c)

akien-mga committed:
PackedScene: Use ObjectID for DeferredNodePathProperties base

Help do proper error handling and avoid a crash.

Works around #103263.

(cherry picked from commit dbdd8a2db922811d8c5975797e44844979801c6d)

Commit:
fix vbox separation in sections with PROPERTY_USAGE_ARRAY

(cherry picked from commit 4952e078e49b3714ffc49b9bfa12d606438a6598)

huwpascoe committed:
Fix 2D instance params crashing using outside of main()

(cherry picked from commit 93bc18f2b0477b30936754b58bd02141774b4336)

dsnopek committed:
JavaClassWrapper: Fix converting returned arrays to Godot types

(cherry picked from commit bbc66056a8465b7b9b608847e32d582de0df9e3c)

berarma committed:
VideoStreamPlayer: stop video on exit tree

(cherry picked from commit bd3a30649bd2be184f61d9a1b2526b351a1bcf52)

Hilderin committed:
Fix headless import always emits errors

(cherry picked from commit dba92c9354c283adcac58408b6ed120a83d93ff1)

darksylinc committed:
Update to latest version of Swappy

Fixes #103294

(cherry picked from commit 89ea5b3d005126a2f7c463f7f487c5001d1b949e)

kitbdev committed:
Fix TextEdit scrolls wrong on text selection

(cherry picked from commit b7a813890766b0103b52b7a0427fdcc37d6c6d50)

m4gr3d committed:
Fix Android mouse capture issues

- Allow mouse capture to be enabled in `_ready`
- Update the input handler logic to avoid dropping mouse captured motion events

(cherry picked from commit bea6472ea479f5c96cad825e71cf3a1d6a69ccdc)

Alex2782 committed:
Fix crash after changing device language

The app was restarting unexpectedly due to missing configChanges flags.
Added "locale|layoutDirection" to AndroidManifest.xml to prevent activity recreation.

(cherry picked from commit 1c02316365a8758af31198a3e0c9d245d3158421)

Chaosus committed:
Fix "unused varying" incorrect warning in shaders

(cherry picked from commit 282497817881fd6dc578a9abeaf070f78ae9acf2)

mihe committed:
Fix broken negative scaling when using Jolt Physics

(cherry picked from commit 62e8b1e5fdeb3ffcd118fbaad68dfc22e721fc00)

arkology committed:
Restore "Show in File Manager" button functionality in `ProjectManager`

- now "Show in File Manager" button is invisible on Android and web (as it should be) when project path is valid.
- tooltip is visible again.

(cherry picked from commit 92c6920272e307d4097517f902429b210ae46c26)

timothyqiu committed:
Set editor's translation domain at root node

(cherry picked from commit 7c8f81f1d07761230ff0471553111f9c2dba8daa)

adamscott committed:
Update UID upgrade tool dialog message

bruvzg committed:
Use separate WorkThreadPool for shader compiler.

(cherry picked from commit 53bb897458a9761d00936ad53ea95bd78b4a6b0b)

KoBeWi committed:
Fix caret_force_displayed

(cherry picked from commit 6d2f18c6c6c9b0415da4d2daed2b75e21a62c350)

akien-mga committed:
Bump version to 4.4.1-rc

dsnopek committed:
CI: Use correct godot-cpp branch

(cherry picked from commit 70bace7482e00e856795b149afe1f3d9ba125f6f)

akien-mga committed:
Android: Fix build with `disable_3d`

Fixes #103516.

(cherry picked from commit a7dd4ad5f87726b87cb7e47020b4a479cb0adadd)

jamie-pate committed:
Fix check for is_maximized in x11 to require both horizontal and vertical

Fixes #103522

Persistent window state doesn't work if your window is 'tiled' in X11
gnome

3.x checks for `horz && vert`
https://github.com/godotengine/godot/blob/4396e98834f159da59ec790f2ff64fb65dacd9ce/platform/x11/os_x11.cpp#L1708

4.x also checked `horz && vert` until this change where it was switched
to `horz || vert`
https://github.com/godotengine/godot/commit/524f061c015b8070de73381bf6b7932514de79de#diff-05f85bc3bf96d384f6b96260c758e63e10bbdd52b04f8ccb34649372e7bc1f48R1382

The corrected logic is:
Check `horz && vert` for 'is_maximized'
Check `horz || vert` for 'can_maximize'

(cherry picked from commit 4f14f722b8826a91b0bf83a3492e20ec4cb352f7)

Hot-key committed:
Fix crash when calling get_argument_count() on invalid Callable

https://github.com/godotengine/godot/issues/103438
(cherry picked from commit 7bc030172e9d90e5c90b32f05624d5cf5070ab6d)

bruvzg committed:
[Linux/BSD] Offload RenderingDevice creation test to subprocess.

(cherry picked from commit 6ed12bfc5d1f8e541134b1440ed51e514490d347)

olukowski committed:
Make NavigationLink3D properly update on NOTIFICATION_VISIBILITY_CHANGED

(cherry picked from commit fc1461cfe0791f0da9828c73c1521d57603636b6)

bruvzg committed:
[iOS export] Restore one-click deploy device enumeration using Xcode.

(cherry picked from commit 4187c48a151039d021edbf3d5dc4a776be8eded9)

TrueJole committed:
Use text_overrun for project export errors

(cherry picked from commit 555caeaecc1315843a30aa1466b6de1f47cb397c)

stuartcarnie committed:
2D: Fix light shader accessing TEXTURE_PIXEL_SIZE

(cherry picked from commit b3694662a26150b04089403f97a5afcf910c3b8f)

bruvzg committed:
[Windows] Fix `get_modified_time` on locked files.

(cherry picked from commit 6597fa24a9eb6a601779b55f39cd0b7f51781cd1)

kleonc committed:
Use Viewport's default texture filter/repeat in GUI tooltips

(cherry picked from commit 91170d0a30295ec9c0dc973c971d72be74614b02)

Grublady committed:
Swap Nintendo face buttons on macOS

(cherry picked from commit cdcee78fcbb4bdb84629e671bf1d57c39fc45937)

pafuent committed:
Fix `SCRIPT ERROR/ERROR/WARNING` on test output

Also "fixing" some GDScript completion tests which
where named wrongly.

(cherry picked from commit 5775d29ad8d72fddc88ee76b7806db2f0568fcf8)

KoBeWi committed:
Update script modified times when saved in EditorNode

(cherry picked from commit dade8bd66fedda9c35ab17146c31861d772c97cd)

BrotherShort committed:
Fix SceneTree's rename LineEdit's offset position

(cherry picked from commit 7d7b773147de91c319d975dc0af9bf4c2faa3363)

raulsntos committed:
[.NET] Skip re-saving `.csproj` when TFM is unchanged

Avoids updating the platform-specific `TargetFramework` properties if they already match the minimum required version.

(cherry picked from commit 57d5b664d35fbb13f1ebec74fa45272b8183e623)

dsnopek committed:
JavaClassWrapper: Fix conversion to/from `org.godotengine.godot.Dictionary` that regressed

(cherry picked from commit f52b2e4db4d785e5f03377de7140808ba99f0f95)

darksylinc committed:
Fix incorrect parameters passed to VMA

If the allocation is small enough that it enters the

if (p_size <= SMALL_ALLOCATION_MAX_SIZE) {} block, Godot would call
vmaFindMemoryTypeIndexForBufferInfo with the wrong parameters.

This can cause vmaFindMemoryTypeIndexForBufferInfo to potentially
misbehave on some cards or drivers.

Fixes regression introduced in #102830
Might potentially reopen #101850 (I doubt it, but it's possible)

Must be backported to 4.4

(cherry picked from commit c543c5615cf2ececbfc60c49e22f74b59851312d)

timothyqiu committed:
Fix some i18n issues in visual shader editor

(cherry picked from commit ef4636ac5966e1ad4ee351f1ea58ca80960b460d)

BlueCube3310 committed:
BasisUniversal: Ensure ASTC's HDR variant is supported when transcoding

(cherry picked from commit 4d46ef8e8e3f8c9cb57902ef18a21ba49b38514d)

AeioMuch committed:
Fix ownership when pasting non root with child nodes

(cherry picked from commit 8d39e4b14ee8c4bf0cfcab4f608c82d47e82858e)

Calinou committed:
Use more efficient sleep approach on Windows when low-processor mode is enabled

This aims to fix the idle CPU utilization regression from 4.3 to 4.4
by reverting to the previous approach, but only when low-processor mode
is enabled.

(cherry picked from commit 03b4e0dd5855971eca0f35ac22d420e9e0900bb6)

stuartcarnie committed:
MetalFX: change fallback behaviour

Closes #103782

(cherry picked from commit f31ddce6d1708d8774262603516afad214e44326)

Rindbee committed:
Fix memory leak caused by hidden tooltip controls

Using hidden controls can override the default tooltip behavior, but these
controls are forgotten to be deleted.

(cherry picked from commit ac19f2ca68e89f53c33c8553775bf1c402d88206)

timothyqiu committed:
ScriptEditor: Disable auto translation of the filename label

(cherry picked from commit f329439249b28e11d6ddab08fccc755a2535aadc)

Bromeon committed:
Register editor-only OpenXR* classes as api_type = "editor" instead of "core"

(cherry picked from commit 99071539e021330987a4668962f6bfb57180bcfb)

mooflu committed:
Fix gles3 gaussian_blur mipmap setup.

Extend max level to include i for writing and so fb is complete and avoid resulting errors like:
"Framebuffer is incomplete: Attachment level is not in the [base level, max level] range".

(cherry picked from commit e9bbe6fc784658f1e7116cb44f3f70e16aae09fd)

shahriarlabib000 committed:
update logo/run icon path in README

(cherry picked from commit 25a8b808d33c8036655b2154bd21aa8fa5f679d4)

timothyqiu committed:
Fix BBCode tag for NativeMenu class reference

(cherry picked from commit 1082b1915641d1352ce2df1de9576e41c6bb80f1)

rburing committed:
CPUParticles2D: Fix fixed timestep interpolation after entering tree with emitting = false

This line was left over from earlier testing; the intended functionality
is now contained in _refresh_interpolation_state().

(cherry picked from commit 5892ddbe24b6938cf5edb0c5d85a88fa12e22883)

dsnopek committed:
OpenXR: Fix OpenGL version warning when using GLES

(cherry picked from commit 63533862416cc21e5193b5dab162680cd7875f90)

bruvzg committed:
[ResourceLoader] Do not wait for the main thread during initial reimport.

(cherry picked from commit 16865b69178f0c653dac07c3713a6088e508e966)

SaracenOne committed:
Fixes missing process_state error in blend spaces.

(cherry picked from commit 250e69856e10ac3290127753caffe4a6d8826606)

timothyqiu committed:
Add missing ETR/TTR markups

(cherry picked from commit a7c1f2730f9ac6891069f1af08b0b564c401b9e9)

YeldhamDev committed:
Fix wrong canvas camera override panning in the runtime debugger

(cherry picked from commit 186d8259d54c0147b63f46ca882b67885b96f782)

ydeltastar committed:
Fix debugger's memory leak when project closes itself

(cherry picked from commit ddb94b9e9d9e6ea5632df863d3f023f678ad1f5c)

adamscott committed:
Set interactive music streams as meta streams

(cherry picked from commit d68ff21fbeb36e66ef49a2c344d7e31d45a81dc7)

akien-mga committed:
Merge pull request #103468 from adamscott/update-uid-upgrade-tool

[4.4] Update UID upgrade tool dialog message

akien-mga committed:
Merge pull request #104019 from akien-mga/4.4-cherrypicks

[4.4] Cherry-picks for the 4.4 branch (future 4.4.1) - 1st batch

bruvzg committed:
Fix `Invalid Task ID` errors in `ResourceLoader`.

(cherry picked from commit d1b63ae16cd89054bfc37f442f18409deec9b969)

Hilderin committed:
Fix missing binding for NOTIFICATION_WM_POSITION_CHANGED

(cherry picked from commit 14b1e40e36d72773c91046615ae6c15652422c77)

JulianHeuser committed:
Copy audio data to dest when bypassing pitch effect

(cherry picked from commit f7bb2c8b1183ffabda11b7b1f938466dbbdec5c4)

akien-mga committed:
Merge pull request #104093 from akien-mga/4.4-cherrypicks

[4.4] Cherry-picks for the 4.4 branch (future 4.4.1) - 2nd batch

akien-mga committed:
i18n: Sync translations with Weblate

beiller committed:
Increment on dq experiment

beiller committed:
Merge remote-tracking branch 'origin/4.4' into HEAD

beiller committed:
checkpoint

Arazalera commented:
It would be useful to select deformation per-bone, for jiggle/squish/deformation bones you would need linear skinning instead

computerologist commented:
Maya's implementation is a good model: per-vertex, there's a weight which linearly blends between the DQ and LBS positions.  This does mean that a mesh which has a blended approach like this has double the skinning cost.

computerologist commented:
> It would be useful to select deformation per-bone, for jiggle/squish/deformation bones you would need linear skinning instead

Maya's implementation is a good model: per-vertex, there's a weight which linearly blends between the DQ and LBS positions. This does mean that a mesh which has a blended approach like this has double the skinning cost.

Arazalera commented:
Only one of the models can be instead used at values 0 and 1
